File: spark-apps/stream_score.scala
// ===========================================================================
// PHASE 11 -- Spark Structured Streaming fraud scorer.   (VELOCITY)
//
// WHY SCALA AND NOT PYSPARK
//   The cluster images disagree on Python: spark-master is Alpine 3.10 with
//   Python 3.7 only, nodemanager is Debian 9 with Python 3.5 only. PySpark
//   refuses to run when driver and executor minor versions differ, and
//   neither image can install the other's version. Scala launches no Python
//   workers at all, so the problem cannot occur. Preprocessing (Phase 4) is
//   already Scala, so this keeps the distributed code in one language.
//
// WHAT IT DOES
//   Watches an HDFS directory. When a file lands, Structured Streaming picks
//   it up as a micro-batch, applies the SAME GBT PipelineModel that Phase 5
//   trained on all 132,500,000 rows, and writes checkpointed verdicts back
//   to HDFS.
//
// HONEST FRAMING
//   Structured Streaming is MICRO-BATCH, not per-event. That is how Spark
//   works, and it is what the report says.
//
// TWO THINGS THAT CATCH PEOPLE OUT, HANDLED HERE
//   1. A streaming file source CANNOT infer schema -- the StructType below is
//      declared explicitly and must match the preprocessed Parquet.
//   2. The saved PipelineModel expects engineered features (per-user amount
//      statistics, MCC frequency) that come from batch aggregates. They are
//      computed ONCE at startup and broadcast into the stream as a
//      stream-static join.
//
// Run:  bash scripts/run_streaming.sh
//       bash scripts/feed_stream.sh     (second terminal)
// ===========================================================================

import java.io.PrintWriter

import org.apache.spark.ml.PipelineModel
import org.apache.spark.ml.functions.vector_to_array
import org.apache.spark.sql.DataFrame
import org.apache.spark.sql.functions._
import org.apache.spark.sql.types._

val HDFS       = "hdfs://namenode:9000"
val BATCH_PATH = HDFS + "/fraudlens/dataset/preprocessed"
val MODEL_PATH = HDFS + "/fraudlens/models/gbt"
val IN_DIR     = HDFS + "/fraudlens/streaming/input"
val OUT_DIR    = HDFS + "/fraudlens/streaming/verdicts"
val CKPT_DIR   = HDFS + "/fraudlens/streaming/checkpoint"
val LOCAL_OUT  = "/artifacts/stream_verdicts"
val TIMEOUT_MS = 30L * 60L * 1000L

spark.conf.set("spark.sql.shuffle.partitions", "8")

println("=" * 66)
println(" PHASE 11: Structured Streaming fraud scorer  (Scala)")
println("=" * 66)

// ---------------------------------------------------------------------------
// Declared schema -- streaming file sources cannot infer one.
// ---------------------------------------------------------------------------
val streamSchema = StructType(Array(
  StructField("user",                        IntegerType,   true),
  StructField("card",                        IntegerType,   true),
  StructField("day",                         IntegerType,   true),
  StructField("hour",                        IntegerType,   true),
  StructField("minute",                      IntegerType,   true),
  StructField("txn_timestamp",               TimestampType, true),
  StructField("day_of_week",                 IntegerType,   true),
  StructField("is_weekend",                  IntegerType,   true),
  StructField("day_of_year",                 IntegerType,   true),
  StructField("amount",                      DoubleType,    true),
  StructField("amount_abs",                  DoubleType,    true),
  StructField("amount_log",                  DoubleType,    true),
  StructField("amount_capped",               DoubleType,    true),
  StructField("is_outlier",                  IntegerType,   true),
  StructField("is_refund",                   IntegerType,   true),
  StructField("use_chip",                    StringType,    true),
  StructField("merchant_name",               StringType,    true),
  StructField("merchant_city",               StringType,    true),
  StructField("merchant_state",              StringType,    true),
  StructField("zip",                         StringType,    true),
  StructField("mcc",                         IntegerType,   true),
  StructField("error_flag",                  IntegerType,   true),
  StructField("error_type",                  StringType,    true),
  StructField("vpn_flag",                    IntegerType,   true),
  StructField("device_ip",                   StringType,    true),
  StructField("device_os",                   StringType,    true),
  StructField("device_lat",                  DoubleType,    true),
  StructField("device_lon",                  DoubleType,    true),
  StructField("geo_missing",                 IntegerType,   true),
  StructField("hw_missing",                  IntegerType,   true),
  StructField("imputed_city",                IntegerType,   true),
  StructField("imputed_state",               IntegerType,   true),
  StructField("imputed_zip",                 IntegerType,   true),
  StructField("merchant_lat",                DoubleType,    true),
  StructField("merchant_lon",                DoubleType,    true),
  StructField("device_merchant_distance_km", DoubleType,    true),
  StructField("is_online",                   IntegerType,   true),
  StructField("is_foreign_or_unknown",       IntegerType,   true),
  StructField("is_night",                    IntegerType,   true),
  StructField("is_fraud",                    IntegerType,   true),
  StructField("processed_at",                TimestampType, true),
  StructField("year",                        IntegerType,   true),
  StructField("month",                       IntegerType,   true)
))

// ---------------------------------------------------------------------------
// Batch-derived features, computed once and broadcast into the stream.
// ---------------------------------------------------------------------------
println(">>> Computing user and MCC statistics from the batch dataset")
val batchDf = spark.read.parquet(BATCH_PATH)

val userStats = batchDf.groupBy("user")
  .agg(avg("amount_abs").as("user_amount_mean"),
       stddev("amount_abs").as("user_amount_std"),
       count("*").as("user_txn_count"))
  .na.fill(Map("user_amount_std" -> 0.0))
  .cache()

val mccFreq = batchDf.groupBy("mcc")
  .agg(count("*").as("mcc_frequency"))
  .cache()

println(f">>> user stats: ${userStats.count()}%,d rows | " +
        f"mcc stats: ${mccFreq.count()}%,d rows")

println(">>> Loading the GBT PipelineModel trained in Phase 5")
val model = PipelineModel.load(MODEL_PATH)
println(">>> Model loaded: " + model.stages.length + " pipeline stages")

// ---------------------------------------------------------------------------
// The stream
// ---------------------------------------------------------------------------
println(">>> Watching " + IN_DIR)
val rawStream = spark.readStream
  .schema(streamSchema)
  .option("header", "true")
  .option("maxFilesPerTrigger", 1)
  .csv(IN_DIR)

val enriched = rawStream
  .join(broadcast(userStats), Seq("user"), "left")
  .join(broadcast(mccFreq), Seq("mcc"), "left")
  .withColumn("amount_vs_user_mean",
    col("amount_abs") / (coalesce(col("user_amount_mean"), lit(0.0)) + lit(1.0)))
  .withColumn("amount_zscore",
    (col("amount_abs") - coalesce(col("user_amount_mean"), lit(0.0))) /
    (coalesce(col("user_amount_std"), lit(0.0)) + lit(1.0)))
  .na.fill(0.0, Array("user_amount_mean", "user_amount_std", "user_txn_count",
                      "amount_vs_user_mean", "amount_zscore", "mcc_frequency"))

val scored = model.transform(enriched)

val verdicts = scored.select(
  col("user"),
  col("amount"),
  col("merchant_state"),
  col("mcc"),
  col("hour"),
  round(vector_to_array(col("probability"))(1), 6).as("fraud_probability"),
  col("prediction").cast(IntegerType).as("predicted_fraud"),
  col("is_fraud").as("actual_fraud"),
  current_timestamp().as("scored_at")
)

// ---------------------------------------------------------------------------
// Sink: persist every verdict to HDFS, mirror a slice locally for the API.
// ---------------------------------------------------------------------------
new java.io.File(LOCAL_OUT).mkdirs()

var totalBatches = 0L
var totalRows    = 0L
var totalFlagged = 0L

val query = verdicts.writeStream
  .foreachBatch { (df: DataFrame, batchId: Long) =>
    val n = df.count()
    if (n > 0) {
      val flagged = df.filter(col("predicted_fraud") === 1).count()
      totalBatches += 1
      totalRows    += n
      totalFlagged += flagged

      df.write.mode("append").parquet(OUT_DIR)

      val sample = df.orderBy(desc("fraud_probability")).limit(25).collect()
      val pw = new PrintWriter(f"$LOCAL_OUT/batch_$batchId%05d.json")
      try {
        sample.foreach { r =>
          pw.println(
            "{\"user\":" + r.getAs[Int]("user") +
            ",\"amount\":" + r.getAs[Double]("amount") +
            ",\"merchant_state\":\"" + r.getAs[String]("merchant_state") + "\"" +
            ",\"mcc\":" + r.getAs[Int]("mcc") +
            ",\"hour\":" + r.getAs[Int]("hour") +
            ",\"fraud_probability\":" + r.getAs[Double]("fraud_probability") +
            ",\"predicted_fraud\":" + r.getAs[Int]("predicted_fraud") +
            ",\"actual_fraud\":" + r.getAs[Int]("actual_fraud") +
            ",\"scored_at\":\"" + r.getAs[java.sql.Timestamp]("scored_at") + "\"}")
        }
      } finally {
        pw.close()
      }

      println(f"[stream] batch $batchId%-4d | $n%6d rows | $flagged%5d flagged " +
              f"| totals: $totalRows%,d rows, $totalFlagged%,d flagged")
    }
  }
  .outputMode("append")
  .option("checkpointLocation", CKPT_DIR)
  .trigger(org.apache.spark.sql.streaming.Trigger.ProcessingTime("5 seconds"))
  .start()

println("")
println("=" * 66)
println(" STREAM RUNNING")
println(" Feed it:  bash scripts/feed_stream.sh    (second terminal)")
println(" Watch it: http://localhost:4040 -> Structured Streaming tab")
println(" Stops automatically after 30 minutes, or press Ctrl-C.")
println("=" * 66)
println("")

query.awaitTermination(TIMEOUT_MS)
query.stop()

println("")
println("=" * 66)
println(f" micro-batches processed : $totalBatches%,d")
println(f" rows scored             : $totalRows%,d")
println(f" flagged as fraud        : $totalFlagged%,d")
println(" verdicts in HDFS        : " + OUT_DIR)
println("=" * 66)

System.exit(0)

File: spark-apps/stream_score.py
#!/usr/bin/env python3
"""
PHASE 11 -- Spark Structured Streaming scorer.  (VELOCITY REQUIREMENT)

WHAT THIS CLOSES
    The course requires Velocity: "streaming data or real-time processing".
    Everything else in this project is batch. This job watches an HDFS
    directory and scores transactions as files arrive, applying the SAME GBT
    PipelineModel that Phase 5 trained on all 132,500,000 rows.

HONEST FRAMING
    Structured Streaming is MICRO-BATCH, not per-event. That is how Spark
    works and it is what will be described in the report.

TWO THINGS THAT CATCH PEOPLE OUT, HANDLED HERE
    1. A streaming file source CANNOT infer schema. The StructType below is
       declared explicitly and must match the preprocessed Parquet schema.
    2. The saved PipelineModel expects the engineered features (user amount
       statistics, MCC frequency). Those come from batch aggregates, so they
       are computed ONCE at startup from the preprocessed dataset and
       broadcast into the stream as a stream-static join.

Run:  bash scripts/run_streaming.sh
      bash scripts/feed_stream.sh     (in a second terminal)
"""
import argparse
import json
import os
import time

from pyspark.ml import PipelineModel
from pyspark.sql import SparkSession
from pyspark.sql import functions as F
from pyspark.sql.types import (DoubleType, IntegerType, StringType,
                               StructField, StructType, TimestampType)

HDFS = "hdfs://namenode:9000"
BATCH = HDFS + "/fraudlens/dataset/preprocessed"
MODEL = HDFS + "/fraudlens/models/gbt"
IN_DIR = HDFS + "/fraudlens/streaming/input"
OUT_DIR = HDFS + "/fraudlens/streaming/verdicts"
CKPT = HDFS + "/fraudlens/streaming/checkpoint"
LOCAL_VERDICTS = "/artifacts/stream_verdicts"

# MUST match the preprocessed schema -- streaming file sources cannot infer it.
SCHEMA = StructType([
    StructField("user", IntegerType(), True),
    StructField("card", IntegerType(), True),
    StructField("day", IntegerType(), True),
    StructField("hour", IntegerType(), True),
    StructField("minute", IntegerType(), True),
    StructField("txn_timestamp", TimestampType(), True),
    StructField("day_of_week", IntegerType(), True),
    StructField("is_weekend", IntegerType(), True),
    StructField("day_of_year", IntegerType(), True),
    StructField("amount", DoubleType(), True),
    StructField("amount_abs", DoubleType(), True),
    StructField("amount_log", DoubleType(), True),
    StructField("amount_capped", DoubleType(), True),
    StructField("is_outlier", IntegerType(), True),
    StructField("is_refund", IntegerType(), True),
    StructField("use_chip", StringType(), True),
    StructField("merchant_name", StringType(), True),
    StructField("merchant_city", StringType(), True),
    StructField("merchant_state", StringType(), True),
    StructField("zip", StringType(), True),
    StructField("mcc", IntegerType(), True),
    StructField("error_flag", IntegerType(), True),
    StructField("error_type", StringType(), True),
    StructField("vpn_flag", IntegerType(), True),
    StructField("device_ip", StringType(), True),
    StructField("device_os", StringType(), True),
    StructField("device_lat", DoubleType(), True),
    StructField("device_lon", DoubleType(), True),
    StructField("geo_missing", IntegerType(), True),
    StructField("hw_missing", IntegerType(), True),
    StructField("imputed_city", IntegerType(), True),
    StructField("imputed_state", IntegerType(), True),
    StructField("imputed_zip", IntegerType(), True),
    StructField("merchant_lat", DoubleType(), True),
    StructField("merchant_lon", DoubleType(), True),
    StructField("device_merchant_distance_km", DoubleType(), True),
    StructField("is_online", IntegerType(), True),
    StructField("is_foreign_or_unknown", IntegerType(), True),
    StructField("is_night", IntegerType(), True),
    StructField("is_fraud", IntegerType(), True),
    StructField("processed_at", TimestampType(), True),
    StructField("year", IntegerType(), True),
    StructField("month", IntegerType(), True),
])

FILL = ["user_amount_mean", "user_amount_std", "user_txn_count",
        "amount_vs_user_mean", "amount_zscore", "mcc_frequency"]


def log(m):
    print("[phase11] " + str(m), flush=True)


def main():
    ap = argparse.ArgumentParser()
    ap.add_argument("--max-files-per-trigger", type=int, default=1)
    ap.add_argument("--timeout-min", type=float, default=30.0)
    args = ap.parse_args()

    spark = (SparkSession.builder
             .appName("FraudLens-Phase11-StreamingScorer")
             .getOrCreate())
    spark.sparkContext.setLogLevel("WARN")

    log("=" * 62)
    log("PHASE 11: Structured Streaming fraud scorer")
    log("=" * 62)

    # ---- batch-derived features, computed once and broadcast -------------
    log("computing user and MCC statistics from the batch dataset")
    batch = spark.read.parquet(BATCH)
    user_stats = (batch.groupBy("user")
                  .agg(F.avg("amount_abs").alias("user_amount_mean"),
                       F.stddev("amount_abs").alias("user_amount_std"),
                       F.count("*").alias("user_txn_count"))
                  .fillna({"user_amount_std": 0.0}))
    mcc_freq = batch.groupBy("mcc").agg(F.count("*").alias("mcc_frequency"))
    user_stats = F.broadcast(user_stats.cache())
    mcc_freq = F.broadcast(mcc_freq.cache())
    log("user stats: %d rows | mcc stats: %d rows"
        % (user_stats.count(), mcc_freq.count()))

    log("loading the GBT PipelineModel trained in Phase 5")
    model = PipelineModel.load(MODEL)

    # ---- the stream -------------------------------------------------------
    log("watching %s" % IN_DIR)
    stream = (spark.readStream
              .schema(SCHEMA)
              .option("header", "true")
              .option("maxFilesPerTrigger", args.max_files_per_trigger)
              .csv(IN_DIR))

    enriched = (stream
                .join(user_stats, on="user", how="left")
                .join(mcc_freq, on="mcc", how="left")
                .withColumn("amount_vs_user_mean",
                            F.col("amount_abs") / (F.col("user_amount_mean") + F.lit(1.0)))
                .withColumn("amount_zscore",
                            (F.col("amount_abs") - F.col("user_amount_mean")) /
                            (F.col("user_amount_std") + F.lit(1.0)))
                .fillna(0.0, subset=FILL))

    scored = model.transform(enriched)

    from pyspark.ml.functions import vector_to_array
    verdicts = scored.select(
        F.col("user"),
        F.col("amount"),
        F.col("merchant_state"),
        F.col("mcc"),
        F.col("hour"),
        F.round(vector_to_array(F.col("probability"))[1], 6).alias("fraud_probability"),
        F.col("prediction").cast(IntegerType()).alias("predicted_fraud"),
        F.col("is_fraud").alias("actual_fraud"),
        F.current_timestamp().alias("scored_at"),
    )

    os.makedirs(LOCAL_VERDICTS, exist_ok=True)
    counter = {"batches": 0, "rows": 0, "flagged": 0}

    def sink(df, batch_id):
        """foreachBatch: persist to HDFS and mirror a slice for the API."""
        rows = df.count()
        if rows == 0:
            return
        flagged = df.filter(F.col("predicted_fraud") == 1).count()
        counter["batches"] += 1
        counter["rows"] += rows
        counter["flagged"] += flagged

        (df.write.mode("append").parquet(OUT_DIR))

        sample = df.orderBy(F.desc("fraud_probability")).limit(25).collect()
        path = os.path.join(LOCAL_VERDICTS, "batch_%05d.json" % batch_id)
        with open(path, "w") as fh:
            for r in sample:
                d = r.asDict()
                d["scored_at"] = str(d.get("scored_at"))
                fh.write(json.dumps(d) + "\n")

        log("batch %-4d | %6d rows | %5d flagged | totals: %d rows, %d flagged"
            % (batch_id, rows, flagged, counter["rows"], counter["flagged"]))

    query = (verdicts.writeStream
             .foreachBatch(sink)
             .outputMode("append")
             .option("checkpointLocation", CKPT)
             .trigger(processingTime="5 seconds")
             .start())

    log("")
    log("STREAM RUNNING. Drip files in with:  bash scripts/feed_stream.sh")
    log("Spark UI -> Structured Streaming tab shows the batch counter live.")
    log("Ctrl-C to stop, or it exits after %.0f minutes." % args.timeout_min)
    log("")

    try:
        query.awaitTermination(timeout=args.timeout_min * 60)
    except KeyboardInterrupt:
        log("interrupted")
    finally:
        query.stop()
        log("=" * 62)
        log("micro-batches processed : %d" % counter["batches"])
        log("rows scored             : %d" % counter["rows"])
        log("flagged as fraud        : %d" % counter["flagged"])
        log("verdicts in HDFS        : %s" % OUT_DIR)
        spark.stop()


if __name__ == "__main__":
    main()
